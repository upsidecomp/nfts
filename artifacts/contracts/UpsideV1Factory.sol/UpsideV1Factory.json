{
  "_format": "hh-sol-artifact-1",
  "contractName": "UpsideV1Factory",
  "sourceName": "contracts/UpsideV1Factory.sol",
  "abi": [
    {
      "inputs": [],
      "stateMutability": "nonpayable",
      "type": "constructor"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": true,
          "internalType": "address",
          "name": "oldOwner",
          "type": "address"
        },
        {
          "indexed": true,
          "internalType": "address",
          "name": "newOwner",
          "type": "address"
        }
      ],
      "name": "OwnerChanged",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": true,
          "internalType": "address",
          "name": "owner",
          "type": "address"
        },
        {
          "indexed": true,
          "internalType": "uint24",
          "name": "fee",
          "type": "uint24"
        },
        {
          "indexed": false,
          "internalType": "address",
          "name": "pool",
          "type": "address"
        }
      ],
      "name": "PoolCreated",
      "type": "event"
    },
    {
      "inputs": [
        {
          "internalType": "address",
          "name": "_owner",
          "type": "address"
        },
        {
          "internalType": "uint24",
          "name": "_fee",
          "type": "uint24"
        }
      ],
      "name": "createPool",
      "outputs": [
        {
          "internalType": "address",
          "name": "pool",
          "type": "address"
        }
      ],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "address",
          "name": "",
          "type": "address"
        }
      ],
      "name": "getPool",
      "outputs": [
        {
          "internalType": "address",
          "name": "",
          "type": "address"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "owner",
      "outputs": [
        {
          "internalType": "address",
          "name": "",
          "type": "address"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "parameters",
      "outputs": [
        {
          "internalType": "address",
          "name": "factory",
          "type": "address"
        },
        {
          "internalType": "address",
          "name": "owner",
          "type": "address"
        },
        {
          "internalType": "uint24",
          "name": "fee",
          "type": "uint24"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "address",
          "name": "_owner",
          "type": "address"
        }
      ],
      "name": "setOwner",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    }
  ],
  "bytecode": "0x608060405234801561001057600080fd5b5033600260006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055503373ffffffffffffffffffffffffffffffffffffffff16600073ffffffffffffffffffffffffffffffffffffffff167fb532073b38c83145e3e5135377a08bf9aab55bc0fd7c1179cd4fb995d2a5159c60405160405180910390a3610d01806100bc6000396000f3fe608060405234801561001057600080fd5b50600436106100575760003560e01c806313af40351461005c57806389035730146100785780638b078beb146100985780638da5cb5b146100c8578063bbe4f6db146100e6575b600080fd5b610076600480360381019061007191906106e2565b610116565b005b610080610230565b60405161008f939291906107c0565b60405180910390f35b6100b260048036038101906100ad919061070b565b610297565b6040516100bf91906107a5565b60405180910390f35b6100d0610485565b6040516100dd91906107a5565b60405180910390f35b61010060048036038101906100fb91906106e2565b6104ab565b60405161010d91906107a5565b60405180910390f35b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff161461017057600080fd5b8073ffffffffffffffffffffffffffffffffffffffff16600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff167fb532073b38c83145e3e5135377a08bf9aab55bc0fd7c1179cd4fb995d2a5159c60405160405180910390a380600260006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050565b60008060000160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16908060010160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16908060010160149054906101000a900462ffffff16905083565b60008073ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff161415610308576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016102ff90610820565b60405180910390fd5b600073ffffffffffffffffffffffffffffffffffffffff16600360008573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16146103a057600080fd5b6103ab3084846104de565b905080600360008573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055508162ffffff168373ffffffffffffffffffffffffffffffffffffffff167ff22a25a14129cccb17bce9e817217f50b87f174703f53d602b376c35e04068818360405161047791906107a5565b60405180910390a392915050565b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b60036020528060005260406000206000915054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b600060405180606001604052808573ffffffffffffffffffffffffffffffffffffffff1681526020018473ffffffffffffffffffffffffffffffffffffffff1681526020018362ffffff168152506000808201518160000160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555060208201518160010160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555060408201518160010160146101000a81548162ffffff021916908362ffffff16021790555090505082826040516020016105f69291906107f7565b60405160208183030381529060405280519060200120604051610618906106ab565b8190604051809103906000f5905080158015610638573d6000803e3d6000fd5b5090506000806000820160006101000a81549073ffffffffffffffffffffffffffffffffffffffff02191690556001820160006101000a81549073ffffffffffffffffffffffffffffffffffffffff02191690556001820160146101000a81549062ffffff021916905550509392505050565b61040b806108c183390190565b6000813590506106c781610892565b92915050565b6000813590506106dc816108a9565b92915050565b6000602082840312156106f457600080fd5b6000610702848285016106b8565b91505092915050565b6000806040838503121561071e57600080fd5b600061072c858286016106b8565b925050602061073d858286016106cd565b9150509250929050565b61075081610851565b82525050565b6000610763601683610840565b91507f55707369646556313a205a45524f5f41444452455353000000000000000000006000830152602082019050919050565b61079f81610883565b82525050565b60006020820190506107ba6000830184610747565b92915050565b60006060820190506107d56000830186610747565b6107e26020830185610747565b6107ef6040830184610796565b949350505050565b600060408201905061080c6000830185610747565b6108196020830184610796565b9392505050565b6000602082019050818103600083015261083981610756565b9050919050565b600082825260208201905092915050565b600061085c82610863565b9050919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b600062ffffff82169050919050565b61089b81610851565b81146108a657600080fd5b50565b6108b281610883565b81146108bd57600080fd5b5056fe60e060405234801561001057600080fd5b503373ffffffffffffffffffffffffffffffffffffffff1663890357306040518163ffffffff1660e01b815260040160606040518083038186803b15801561005757600080fd5b505afa15801561006b573d6000803e3d6000fd5b505050506040513d601f19601f8201168201806040525081019061008f9190610141565b8062ffffff1660c09062ffffff1660e81b8152508173ffffffffffffffffffffffffffffffffffffffff1660a09073ffffffffffffffffffffffffffffffffffffffff1660601b8152508273ffffffffffffffffffffffffffffffffffffffff1660809073ffffffffffffffffffffffffffffffffffffffff1660601b8152505050506101ff565b600081519050610126816101d1565b92915050565b60008151905061013b816101e8565b92915050565b60008060006060848603121561015657600080fd5b600061016486828701610117565b935050602061017586828701610117565b92505060406101868682870161012c565b9150509250925092565b600061019b826101a2565b9050919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b600062ffffff82169050919050565b6101da81610190565b81146101e557600080fd5b50565b6101f1816101c2565b81146101fc57600080fd5b50565b60805160601c60a05160601c60c05160e81c6101d7610234600039600060ea0152600060a20152600060c601526101d76000f3fe608060405234801561001057600080fd5b50600436106100415760003560e01c80638da5cb5b14610046578063c45a015514610064578063ddca3f4314610082575b600080fd5b61004e6100a0565b60405161005b919061012a565b60405180910390f35b61006c6100c4565b604051610079919061012a565b60405180910390f35b61008a6100e8565b6040516100979190610145565b60405180910390f35b7f000000000000000000000000000000000000000000000000000000000000000081565b7f000000000000000000000000000000000000000000000000000000000000000081565b7f000000000000000000000000000000000000000000000000000000000000000081565b61011581610160565b82525050565b61012481610192565b82525050565b600060208201905061013f600083018461010c565b92915050565b600060208201905061015a600083018461011b565b92915050565b600061016b82610172565b9050919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b600062ffffff8216905091905056fea2646970667358221220b2145763497ea6c3213cb2517554a3b9787d6a2d9a3a4926ae3e221e2cc7e1df64736f6c63430008000033a264697066735822122084688657987c45f8dc1da98597ca43037c09f1b960518677f9ed8ee8be66da9764736f6c63430008000033",
  "deployedBytecode": "0x608060405234801561001057600080fd5b50600436106100575760003560e01c806313af40351461005c57806389035730146100785780638b078beb146100985780638da5cb5b146100c8578063bbe4f6db146100e6575b600080fd5b610076600480360381019061007191906106e2565b610116565b005b610080610230565b60405161008f939291906107c0565b60405180910390f35b6100b260048036038101906100ad919061070b565b610297565b6040516100bf91906107a5565b60405180910390f35b6100d0610485565b6040516100dd91906107a5565b60405180910390f35b61010060048036038101906100fb91906106e2565b6104ab565b60405161010d91906107a5565b60405180910390f35b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff161461017057600080fd5b8073ffffffffffffffffffffffffffffffffffffffff16600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff167fb532073b38c83145e3e5135377a08bf9aab55bc0fd7c1179cd4fb995d2a5159c60405160405180910390a380600260006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050565b60008060000160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16908060010160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16908060010160149054906101000a900462ffffff16905083565b60008073ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff161415610308576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016102ff90610820565b60405180910390fd5b600073ffffffffffffffffffffffffffffffffffffffff16600360008573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16146103a057600080fd5b6103ab3084846104de565b905080600360008573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055508162ffffff168373ffffffffffffffffffffffffffffffffffffffff167ff22a25a14129cccb17bce9e817217f50b87f174703f53d602b376c35e04068818360405161047791906107a5565b60405180910390a392915050565b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b60036020528060005260406000206000915054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b600060405180606001604052808573ffffffffffffffffffffffffffffffffffffffff1681526020018473ffffffffffffffffffffffffffffffffffffffff1681526020018362ffffff168152506000808201518160000160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555060208201518160010160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555060408201518160010160146101000a81548162ffffff021916908362ffffff16021790555090505082826040516020016105f69291906107f7565b60405160208183030381529060405280519060200120604051610618906106ab565b8190604051809103906000f5905080158015610638573d6000803e3d6000fd5b5090506000806000820160006101000a81549073ffffffffffffffffffffffffffffffffffffffff02191690556001820160006101000a81549073ffffffffffffffffffffffffffffffffffffffff02191690556001820160146101000a81549062ffffff021916905550509392505050565b61040b806108c183390190565b6000813590506106c781610892565b92915050565b6000813590506106dc816108a9565b92915050565b6000602082840312156106f457600080fd5b6000610702848285016106b8565b91505092915050565b6000806040838503121561071e57600080fd5b600061072c858286016106b8565b925050602061073d858286016106cd565b9150509250929050565b61075081610851565b82525050565b6000610763601683610840565b91507f55707369646556313a205a45524f5f41444452455353000000000000000000006000830152602082019050919050565b61079f81610883565b82525050565b60006020820190506107ba6000830184610747565b92915050565b60006060820190506107d56000830186610747565b6107e26020830185610747565b6107ef6040830184610796565b949350505050565b600060408201905061080c6000830185610747565b6108196020830184610796565b9392505050565b6000602082019050818103600083015261083981610756565b9050919050565b600082825260208201905092915050565b600061085c82610863565b9050919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b600062ffffff82169050919050565b61089b81610851565b81146108a657600080fd5b50565b6108b281610883565b81146108bd57600080fd5b5056fe60e060405234801561001057600080fd5b503373ffffffffffffffffffffffffffffffffffffffff1663890357306040518163ffffffff1660e01b815260040160606040518083038186803b15801561005757600080fd5b505afa15801561006b573d6000803e3d6000fd5b505050506040513d601f19601f8201168201806040525081019061008f9190610141565b8062ffffff1660c09062ffffff1660e81b8152508173ffffffffffffffffffffffffffffffffffffffff1660a09073ffffffffffffffffffffffffffffffffffffffff1660601b8152508273ffffffffffffffffffffffffffffffffffffffff1660809073ffffffffffffffffffffffffffffffffffffffff1660601b8152505050506101ff565b600081519050610126816101d1565b92915050565b60008151905061013b816101e8565b92915050565b60008060006060848603121561015657600080fd5b600061016486828701610117565b935050602061017586828701610117565b92505060406101868682870161012c565b9150509250925092565b600061019b826101a2565b9050919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b600062ffffff82169050919050565b6101da81610190565b81146101e557600080fd5b50565b6101f1816101c2565b81146101fc57600080fd5b50565b60805160601c60a05160601c60c05160e81c6101d7610234600039600060ea0152600060a20152600060c601526101d76000f3fe608060405234801561001057600080fd5b50600436106100415760003560e01c80638da5cb5b14610046578063c45a015514610064578063ddca3f4314610082575b600080fd5b61004e6100a0565b60405161005b919061012a565b60405180910390f35b61006c6100c4565b604051610079919061012a565b60405180910390f35b61008a6100e8565b6040516100979190610145565b60405180910390f35b7f000000000000000000000000000000000000000000000000000000000000000081565b7f000000000000000000000000000000000000000000000000000000000000000081565b7f000000000000000000000000000000000000000000000000000000000000000081565b61011581610160565b82525050565b61012481610192565b82525050565b600060208201905061013f600083018461010c565b92915050565b600060208201905061015a600083018461011b565b92915050565b600061016b82610172565b9050919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b600062ffffff8216905091905056fea2646970667358221220b2145763497ea6c3213cb2517554a3b9787d6a2d9a3a4926ae3e221e2cc7e1df64736f6c63430008000033a264697066735822122084688657987c45f8dc1da98597ca43037c09f1b960518677f9ed8ee8be66da9764736f6c63430008000033",
  "linkReferences": {},
  "deployedLinkReferences": {}
}
