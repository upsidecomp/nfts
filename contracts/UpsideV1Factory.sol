// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

import "./interfaces/IUpsideV1Factory.sol";

import "./UpsideV1PoolDeployer.sol";

import "./UpsideV1Pool.sol";

/// @title Canonical Uniswap V3 factory
/// @notice Deploys Uniswap V3 pools and manages ownership and control over pool protocol fees
contract UpsideV1Factory is IUpsideV1Factory, UpsideV1PoolDeployer {
    /// @inheritdoc IUpsideV1Factory
    address public override owner;

    /// @inheritdoc IUpsideV1Factory
    mapping(address => address) public override getPool;

    constructor() {
        owner = msg.sender;
        emit OwnerChanged(address(0), msg.sender);
    }

    /// @inheritdoc IUpsideV1Factory
    function createPool(address _owner, uint24 _fee) external override returns (address pool) {
        require(_owner != address(0), "UpsideV1: ZERO_ADDRESS");
        require(getPool[_owner] == address(0));

        pool = deploy(address(this), _owner, _fee);
        getPool[_owner] = pool;

        emit PoolCreated(_owner, _fee, pool);
    }

    /// @inheritdoc IUpsideV1Factory
    function setOwner(address _owner) external override {
        require(msg.sender == owner);
        emit OwnerChanged(owner, _owner);
        owner = _owner;
    }
}
