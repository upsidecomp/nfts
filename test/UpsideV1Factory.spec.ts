import { ethers, waffle } from 'hardhat'
import { UpsideV1Factory } from '../typechain/UpsideV1Factory'
import { expect } from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'

import { FeeAmount, getCreate2Address } from './shared/utilities'

const { constants } = ethers

const createFixtureLoader = waffle.createFixtureLoader

const POOL_OWNER = '0x1000000000000000000000000000000000000000'

describe('UpsideV1Factory', () => {
  const [wallet, other] = waffle.provider.getWallets()

  let factory: UpsideV1Factory
  let poolBytecode: string

  const fixture = async () => {
    const factoryFactory = await ethers.getContractFactory('UpsideV1Factory')
    return (await factoryFactory.deploy()) as UpsideV1Factory
  }

  let loadFixture: ReturnType<typeof createFixtureLoader>
  before('create fixture loader', async () => {
    loadFixture = createFixtureLoader([wallet, other])
  })

  before('load pool bytecode', async () => {
    poolBytecode = (await ethers.getContractFactory('UpsideV1Pool')).bytecode
  })

  beforeEach('deploy factory', async () => {
    factory = await loadFixture(fixture)
  })

  it('owner is deployer', async () => {
    expect(await factory.owner()).to.eq(wallet.address)
  })

  it('factory bytecode size', async () => {
    expect(((await waffle.provider.getCode(factory.address)).length - 2) / 2).to.matchSnapshot()
  })

  it('pool bytecode size', async () => {
    await factory.createPool(FeeAmount.MEDIUM)
    const poolAddress = getCreate2Address(factory.address, FeeAmount.MEDIUM, poolBytecode)
    expect(((await waffle.provider.getCode(poolAddress)).length - 2) / 2).to.matchSnapshot()
  })

  async function createAndCheckPool(feeAmount: FeeAmount) {
    const create2Address = getCreate2Address(factory.address, feeAmount, poolBytecode)

    const create = factory.createPool(feeAmount)
    const poolContractFactory = await ethers.getContractFactory('UpsideV1Pool')
    const pool = poolContractFactory.attach(create2Address)

    await expect(create).to.emit(factory, 'PoolCreated').withArgs(feeAmount, create2Address)
    await expect(factory.createPool(feeAmount)).to.be.reverted
    expect(await factory.pool(), 'pool').to.eq(create2Address)
    expect(await pool.factory(), 'pool factory address').to.eq(factory.address)
    expect(await pool.feePercentage(), 'pool fee').to.eq(feeAmount)
  }

  describe('#createPool', () => {
    it('succeeds for low fee pool', async () => {
      await createAndCheckPool(FeeAmount.LOW)
    })

    it('succeeds for medium fee pool', async () => {
      await createAndCheckPool(FeeAmount.MEDIUM)
    })
    it('succeeds for high fee pool', async () => {
      await createAndCheckPool(FeeAmount.HIGH)
    })

    // todo: fix
    // it("fails if fee amount is not enabled", async () => {
    //   await expect(factory.createPool(POOL_OWNER, 250)).to.be.reverted;
    // });

    it('gas', async () => {
      await snapshotGasCost(factory.createPool(FeeAmount.MEDIUM))
    })
  })

  describe('#setOwner', () => {
    it('fails if caller is not owner', async () => {
      await expect(factory.connect(other).setOwner(wallet.address)).to.be.reverted
    })

    it('updates owner', async () => {
      await factory.setOwner(other.address)
      expect(await factory.owner()).to.eq(other.address)
    })

    it('emits event', async () => {
      await expect(factory.setOwner(other.address))
        .to.emit(factory, 'OwnerChanged')
        .withArgs(wallet.address, other.address)
    })

    it('cannot be called by original owner', async () => {
      await factory.setOwner(other.address)
      await expect(factory.setOwner(wallet.address)).to.be.reverted
    })
  })
})
