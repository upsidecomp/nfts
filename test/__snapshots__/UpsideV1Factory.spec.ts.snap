// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpsideV1Factory #createPool gas 1`] = `159175`;

exports[`UpsideV1Factory factory bytecode size 1`] = `1619`;

exports[`UpsideV1Factory pool bytecode size 1`] = `247`;
